### PR TITLE
[Render][Speculator] Add return_loss_mask to render endpoint for training data generation

### PR DESCRIPTION
**Pull request by WindChimeRan**
## Purpose

Adds a `return_loss_mask` field to `ChatCompletionRequest` and a `loss_mask` field to `GenerateRequest`, so `POST /v1/chat/completions/render` can return a per-token assistant/trainable mask alongside `token_ids`. This is needed by downstream training frameworks (e.g. [[speculators](https://github.com/vllm-project/speculators/issues/652)](https://github.com/vllm-project/speculators/issues/652)) that use the render endpoint to delegate tokenization to vLLM instead of duplicating it locally with HF `apply_chat_template`. The mask is computed in a single `apply_chat_template` call (no second pass) via HF's `return_assistant_tokens_mask` when the chat template has `{% generation %}` tags; otherwise `loss_mask` is `null` and the caller handles the fallback. 
## Test

* added test on the new behavior of return_assistant_tokens_mask. No breaking change before / after this PR
* end to end test with speculator. Speculator works with the new renderer endpoint. https://github.com/vllm-project/speculators/pull/665

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by benchislett** [comment]
Why deviate from the established naming convention of "return_assistant_tokens_mask"? I think this is more aligned with the functionality this service would provide, rather than "return_loss_mask" which just reflects what a training server would want to do with it

File: tests/entrypoints/serve/render/test_render.py [hunk 1]
```diff
@@ -369,3 +369,79 @@ async def test_completion_render_multiple_prompts_token_offsets(client):
...
+
+    # The model's template likely doesn't support
+    # return_assistant_tokens_mask, so loss_mask may be null.
+    # If it does return a mask, validate its structure.
+    loss_mask = data.get("loss_mask")
+    if loss_mask is not None:
+        token_ids = data["token_ids"]
+        assert isinstance(loss_mask, list)
+        assert len(loss_mask) == len(token_ids)
+        assert all(v in (0, 1) for v in loss_mask)
```
Comment (benchislett) [inline tests/entrypoints/serve/render/test_render.py]: Please add an actual assertion here that the masked tokens, when detokenized, produce the expected output string. I would like to ensure that the correctness of the loss mask be enforced, e.g. in case a bug occurs where loss_mask returns only zeros

File: vllm/renderers/hf.py [hunk 2]
```diff
@@ -737,23 +781,23 @@ def safe_apply_chat_template(
...
     except Exception as e:
-        # Log and report any library-related exceptions for further
-        # investigation.
         logger.exception(
             "An error occurred in `transformers` while applying chat template"
         )
         raise ValueError(str(e)) from e
 
+    if return_loss_mask:
+        return list(plain), None  # type: ignore[arg-type]
```
Comment (benchislett) [inline vllm/renderers/hf.py]: Is `list(plain)` just a type consistency thing? Or would it actually convert a `str` into a `list`? Maybe it's better to just do `assert isinstance(plain, list)` to satisfy the type checker?

Or is it possible that `plain` is like a tensor or numpy array here?

**Comment by benchislett** [comment]
Besides the naming, this generally LGTM. Ideally would like a review from a codeowner here though

File: vllm/renderers/params.py [hunk 3]
```diff
@@ -87,6 +87,9 @@ class ChatParams:
     mm_processor_kwargs: dict[str, Any] | None = None
     """The kwargs to pass to the multi-modal processor."""
 
+    return_assistant_tokens_mask: bool = False
+    """Request a per-token assistant/loss mask from apply_chat_template."""
```
Comment (benchislett) [inline vllm/renderers/params.py]:
```suggestion
    """Request a per-token assistant mask from apply_chat_template."""
```

File: tests/entrypoints/serve/render/test_render.py [hunk 4]
```diff
@@ -369,3 +369,126 @@ async def test_completion_render_multiple_prompts_token_offsets(client):
...
+    )
+    assert detok.status_code == 200
+    assert "Hi!" in detok.json()["prompt"]
+
+    detok_rest = await client.post(
+        "/detokenize",
+        json={"model": MODEL_NAME, "tokens": unmasked_ids},
+    )
+    assert detok_rest.status_code == 200
+    assert "Hi!" not in detok_rest.json()["prompt"]
```
Comment (benchislett) [inline tests/entrypoints/serve/render/test_render.py:494]:
```suggestion
    assert "Hi!" not in detok_rest.json()["prompt"]
    assert "Bye" in detok_rest.json()["prompt"]
```

File: vllm/renderers/hf.py [hunk 5]
```diff
@@ -1150,6 +1235,8 @@ async def _process_tokens_async(
                 tensors,
                 mm_updates,
             )
+        if assistant_tokens_mask is not None:
```
Comment (DarkLight1337) [inline vllm/renderers/hf.py:1238]: Let's add a new property to `TokensInput` and `MultiModalInput` to document it properly and avoid the hacky `cast` operation

**Review by DarkLight1337 — APPROVED** [review]
